feat: Model Transformation Layer — Native PipelineStage modifiers (#410)

JesperDramsch opened the pull request:
**Part of the checkpoint system refactor (#248).** Implements the Model Transformation Layer as native `PipelineStage` classes built on Phase 1 infrastructure (PR #501).

> ⚠️ **Architecture change from original PR.** The original `train/modify.py` (sync `ModelModifier.apply(model)` interface) has been **deleted and replaced** with native async `PipelineStage` implementations in `checkpoint/modifiers/`. See design decisions below.

### What this PR does

- **`FreezingModifierStage`** — native `PipelineStage` that freezes specified submodules. Full feature port from the original `FreezingModelModifier`: strict mode, gradient validation, parameter counting, dot-notation submodule lookup.
- **Deletes `train/modify.py`** — `ModelModifierApplier`, `FreezingModelModifier`, and `TransferLearningModelModifier` are all removed. Transfer learning now lives in the loading layer (#998). Pipeline handles orchestration.
- **Config templates** for all 4 loading strategies (`weights_only.yaml`, `transfer_learning.yaml`, `warm_start.yaml`, `cold_start.yaml`)
- **Package exports** and catalog discovery for the modifier layer

### Architecture

```
CheckpointContext enters with model (weights already loaded by loading strategy)
        ↓
  FreezingModifierStage freezes specified submodules
        ↓
CheckpointContext exits with model modified + metadata["modifiers_applied"] updated
```

Composes naturally with sources and loaders:
```python
pipeline = CheckpointPipeline(stages=[
    LocalSource(),
    TransferLearningLoader(skip_mismatched=True),
    FreezingModifierStage(submodules_to_freeze=["encoder"]),
])
```

### Why `train/modify.py` was deleted (not wrapped)

The original plan was thin adapter wrappers around the existing sync classes. That turned out to be wrong:

- PR #410's code was prototype, not production-ready
- Wrapping sync `apply(model)` inside async `process(context)` creates dual code paths
- `ModelModifierApplier` duplicated pipeline orchestration that Phase 1 already handles
- `TransferLearningModelModifier` mixed loading + modification concerns — these are now separate layers

Building native `PipelineStage` classes is cleaner: no indirection, single code path, clear separation of concerns.

### Current status

🟢 **Ready for review.** `main` is merged in, so the diff is now scoped to the modifier delta.

- ✅ g1: FreezingModifierStage native port + train/modify.py deleted
- ✅ g2: Config templates
- ✅ g3: Package exports + catalog discovery (ModelModifier ABC + catalog registration)
- ✅ g4: Modifier + discovery tests green (242 checkpoint unit tests pass)
- 🟢 g5: Submitted for review

The catalog-discovery suite also checks that the loading strategies import; that
single check depends on #998 landing on `main` and is out of scope for this PR.

### What's NOT in this PR

| Deferred | Why | Tracked in |
|----------|-----|-----------|
| PEFT/LoRA modifiers | Phase 3 roadmap | #248 |
| QLoRA, Prefix Tuning | Phase 3 roadmap | #248 |
| `train.py` wiring | Phase 3 integration | #495 |

### Related PRs

- **#501** — Phase 1 Pipeline Infrastructure (merged, dependency)
- **#464** — Checkpoint Sources (in review)
- **#998** — Loading Orchestration (in review)
- **#495** — Phase 3 Integration (blocked on Phase 2)

**Part of #248**

<details>
<summary>Original PR description (archived 2026-03-20)</summary>

## Overview
During my work towards #248 I have converged on a solution that requires a refactor of the model initialisation.

This issue implements the **Model Transformation Layer** that applies post-loading modifications to models.

### Components
- `FreezingModelModifier`: Selective parameter freezing by module name
- `TransferLearningModelModifier`: Transfer learning with fallback for missing checkpoint system
- `ModelModifierApplier`: Orchestrates application of multiple modifiers in sequence

### Configuration
```yaml
training:
  model_modifier:
    modifiers:
      - _target_: "anemoi.training.train.modify.FreezingModelModifier"
        submodules_to_freeze:
          - "encoder"
          - "processor.0"
```

</details>



<!-- readthedocs-preview anemoi-training start -->
----
📚 Documentation preview 📚: https://anemoi-training--442.org.readthedocs.build/en/442/

<!-- readthedocs-preview anemoi-training end -->

<!-- readthedocs-preview anemoi-graphs start -->
----
📚 Documentation preview 📚: https://anemoi-graphs--442.org.readthedocs.build/en/442/

<!-- readthedocs-preview anemoi-graphs end -->

<!-- readthedocs-preview anemoi-models start -->
----
📚 Documentation preview 📚: https://anemoi-models--442.org.readthedocs.build/en/442/

<!-- readthedocs-preview anemoi-models end -->


<!-- readthedocs-preview anemoi-training start -->
----
📚 Documentation preview 📚: https://anemoi-training--442.org.readthedocs.build/en/442/

<!-- readthedocs-preview anemoi-training end -->

<!-- readthedocs-preview anemoi-graphs start -->
----
📚 Documentation preview 📚: https://anemoi-graphs--442.org.readthedocs.build/en/442/

<!-- readthedocs-preview anemoi-graphs end -->

<!-- readthedocs-preview anemoi-models start -->
----
📚 Documentation preview 📚: https://anemoi-models--442.org.readthedocs.build/en/442/

<!-- readthedocs-preview anemoi-models end -->

JesperDramsch committed:
feat: model modifiers [skip ci]

JesperDramsch committed:
refactor: adjust configs for model_modifiers

JesperDramsch committed:
feat: implement model_modifier_applier

JesperDramsch committed:
refactor: simplify interfaces for model modifiers

JPXKQX commented:
Thanks Jesper, I think this refactoring makes a lot of sense. Would it make sense to have another modifier "ResumeRun..." to which we could bring all the logic from  `run_id`, `fork_run_id`, `load_only_weights` and `warm_start`?

JesperDramsch commented:
> Thanks Jesper, I think this refactoring makes a lot of sense. Would it make sense to have another modifier "ResumeRun..." to which we could bring all the logic from `run_id`, `fork_run_id`, `load_only_weights` and `warm_start`?

Possibly. I believe my original design around `fork_run_id` ended up confusing most people, so we could take a look whether this design could be used to fix that.

JesperDramsch committed:
feat: integrate checkpoint loading infrastructure from #458

Add extensible checkpoint loading system that separates checkpoint
source handling from model weight loading strategies.

Changes:
- Add CheckpointLoaderRegistry for extensible source loading
  (local, S3, HTTP, GCS, Azure support)
- Add ModelLoaderRegistry for weight loading strategies
  (standard, weights_only, transfer_learning)
- Implement registry pattern for future extensibility

This infrastructure enables:
- Remote checkpoint loading from cloud storage
- Modular loading strategies
- Clean separation of concerns
- Foundation for advanced features (quantization, PEFT)

Depends on: #458 branch infrastructure
Related: #422, #410

JesperDramsch committed:
refactor: migrate model modifiers to use checkpoint loading system

Replace WeightsInitModelModifier with new checkpoint loading architecture
and integrate checkpoint loading into training pipeline.

Changes:
- Remove WeightsInitModelModifier (functionality moved to checkpoint_loading)
- Update TransferLearningModelModifier to use new model_loading system
- Add configurable strict and skip_mismatched parameters
- Integrate checkpoint loading in training pipeline before model modifiers
- Add _load_checkpoint_if_configured method to trainer

Benefits:
- Clean separation: checkpoint loading vs model transformation
- Better parameter control for transfer learning
- DRY principle: single checkpoint loading implementation
- Extensible: prepare for quantization, PEFT features

Related: #422, #410
Depends: checkpoint loading infrastructure

JesperDramsch committed:
feat: add checkpoint loading configuration system

Introduce new configuration schema and templates for the checkpoint
loading system to replace legacy WeightsInitModelModifier configs.

Changes:
- Add checkpoint_loading field to training schema with Pydantic validation
- Create checkpoint_loading config directory with templates:
  * weights_only.yml - Load only model weights
  * transfer_learning.yml - Load with size mismatch handling
  * standard.yml - Full Lightning checkpoint loading
- Update transfer_learning.yml with new parameters (strict, skip_mismatched)
- Add enhanced_fine_tuning.yml example combining transfer learning + freezing

Benefits:
- Schema validation for checkpoint loading configurations
- Pre-built templates for common use cases
- Flexible parameter configuration per loader type
- Clear separation from model modifier configs

Related: #422, #410

JesperDramsch committed:
test: update and add comprehensive tests for checkpoint loading

Refactor existing tests for new architecture and add comprehensive
test coverage for the checkpoint loading system.

Changes:
- Remove WeightsInitModelModifier tests (functionality moved)
- Update TransferLearningModelModifier tests for new #458 integration
- Add checkpoint loading integration tests for training pipeline
- Add comprehensive test suites from #458 branch:
  * test_checkpoint_loaders.py - Source loading tests (S3, HTTP, etc.)
  * test_model_loading.py - Weight loading strategy tests
- Update integration tests for new model modifier workflow
- Add configuration validation and error handling tests

Coverage:
- All loader types (weights_only, transfer_learning, standard)
- Remote checkpoint sources (S3, HTTP, GCS, Azure)
- Training pipeline integration
- Configuration validation and error scenarios
- Model modifier compatibility with new system

Related: #422, #410

JesperDramsch committed:
Update documentation and fix compatibility issues

- Update documentation with comprehensive checkpoint loading sections
- Fix compatibility issues in forecaster and checkpoint utilities
- Update config templates to maintain backward compatibility
- Fix pre-commit hook issues (ruff RET504, docsig parameter mismatches)

JesperDramsch committed:
wip: End of Day commit

JesperDramsch committed:
docs: improve modelmodifier documentation

JesperDramsch committed:
fix: sort out #458 and #410

JesperDramsch committed:
feat(model-modifiers): enhance freezing with gradient validation

- Add gradient validation during freezing to ensure parameters are truly frozen
- Implement optimized module lookup using PyTorch's get_submodule()
- Improve error messages with clear context for debugging
- Add comprehensive module-level documentation
- Enhance test coverage with improved documentation
- Add noqa comments for necessary complex methods

The gradient validation ensures frozen parameters don't accumulate gradients during
training, providing runtime verification of the freezing mechanism. The optimized
lookup improves performance for deep models by using O(1) access when possible.

JesperDramsch committed:
refactor(model-modifiers): remove TransferLearningModelModifier

Transfer learning checkpoint loading is the responsibility of the
checkpoint acquisition layer (PR #458), not the model modification
layer. This separation of concerns keeps model modifiers focused on
post-loading transformations like parameter freezing.

JesperDramsch committed:
chore(config): remove transfer learning config templates

These configs referenced TransferLearningModelModifier which has been
moved to the checkpoint acquisition layer. The freezing config remains
as the primary model_modifier configuration template.

JesperDramsch committed:
Merge origin/main into PR #410 branch

Resolve 9 merge conflicts. Keep model_modifier system from PR #410
while accepting main's new features: update_ds_stats_on_ckpt_load,
ensemble_size_per_device, recompile_limit, warm_start input,
fit_parameters property, _validate_transfer_learning_datasets.

JesperDramsch committed:
checkpoint: modifiers-g1 — FreezingModifierStage native port (#410)

JesperDramsch committed:
checkpoint: modifiers-g1 cleanup — delete train/modify.py, fix imports and docs (#410)

- Remove train/modify.py and its tests (replaced by checkpoint/modifiers/freezing.py)
- Remove stale ModelModifierApplier import and usage from train.py
- Update docs to reference FreezingModifierStage with new _target_ path
- Remove references to TransferLearningModelModifier (never existed as real class)

JesperDramsch committed:
checkpoint: merge main into modifier branch — resolve schema/config/doc conflicts (#410)

Resolves three content conflicts from merging origin/main:
- schemas/training.py: drop the dangling task docstring (HEAD cleanup), keep the model_modifier field.
- config/training/single.yaml: main renamed default.yaml -> single.yaml; keep main's defaults plus 'optional model_modifier: null'.
- docs/user-guide/training.rst: keep the FreezingModifierStage example and freezing note, append main's Precision/BLAS and Weight Averaging sections.

Relocate tests/checkpoint/modifiers/ -> tests/unit/checkpoint/modifiers/ to match main's test layout.

JesperDramsch committed:
checkpoint: recreate freezing.yml as FreezingModifierStage template (#410)

The modifiers-g1 cleanup deleted the old freezing.yml (it targeted the removed
train.modify.FreezingModelModifier). Recreate it as a single-object template for
the native FreezingModifierStage pipeline stage, with keys mapping to its
constructor (submodules_to_freeze, strict, validate_gradients).

JesperDramsch committed:
checkpoint: modifiers-g2 — config templates for loading strategies (#410)

Add Hydra config templates for the four checkpoint loading strategies
(WeightsOnlyLoader, ColdStartLoader, TransferLearningLoader, WarmStartLoader),
each a single instantiable object (top-level _target_) rather than a
list-under-key shape, since loaders are mutually exclusive — so
hydra.utils.instantiate(cfg) works without unwrapping. Each carries a header
comment stating what the loader does and that it does not change architecture.

JesperDramsch committed:
Merge remote-tracking branch 'origin/main' into 410-refactor-of-model-initialisation-ie-weight-loading-model-freezing-transfer-learning

# Conflicts:
#	training/src/anemoi/training/utils/checkpoint.py

JesperDramsch committed:
checkpoint: modifiers-g3 — ModelModifier ABC + package exports (#410)

Adds the thin `ModelModifier(PipelineStage)` ABC under
`checkpoint/modifiers/base.py`, mirroring the established
`CheckpointSource` / `LoadingStrategy` pattern. The frozen
`ComponentCatalog` scans the modifiers package looking for a base
class named `ModelModifier`; without this layer, discovery returned
an empty registry even though `FreezingModifierStage` was importable.

`FreezingModifierStage` is re-parented from `PipelineStage` to
`ModelModifier`; the g1 `issubclass(..., PipelineStage)` invariant
holds transitively. `modifiers/__init__.py` exports both classes and
gains the standard copyright header. `CheckpointContext` is moved
under `TYPE_CHECKING` to satisfy ruff TC001 on the touched file.

After this change:

  ComponentCatalog.list_modifiers()
  # -> ['freezing_modifier_stage']

JesperDramsch committed:
test(checkpoint): add modifier discovery tests and expand freezing coverage

Add test_discovery.py covering ModelModifier base class, FreezingModifierStage
subclassing, and ComponentCatalog discovery/target resolution for modifiers. An
autouse fixture resets the catalog's class-level discovery cache so the tests are
robust to ordering against test_catalog.py, which leaves the cache populated with
mocked empty results.

Expand test_freezing.py to exercise the full FreezingModifierStage feature set:
strict-mode raising, non-strict skipping, dot-notation submodule paths, gradient
validation, and parameter-count metadata.

JesperDramsch committed:
chore(training): remove stale train/forecaster module resurrected by main merge

JesperDramsch committed:
fix(checkpoint): accumulate modifiers_applied metadata across modifiers

JesperDramsch committed:
fix(config): make model_modifier template a composable modifiers list

JesperDramsch committed:
Merge branch 'main' into 410-refactor-of-model-initialisation-ie-weight-loading-model-freezing-transfer-learning

JesperDramsch committed:
Merge remote-tracking branch 'origin/main' into 410-refactor-of-model-initialisation-ie-weight-loading-model-freezing-transfer-learning

# Conflicts:
#	training/docs/user-guide/training.rst

JesperDramsch committed:
fix(training): align FreezingModifierStage path semantics with #1159

The legacy freeze_submodule_by_name switched to full-path matching in
#1159: an exact child name or dot-separated path, no name-match search
at arbitrary depth, and a found-flag so the caller can warn and skip.

Align the modifier stage with those semantics:
- resolve targets via get_submodule only; remove the recursive
  any-depth fallback (now divergent from legacy)
- distinguish 'path not found' (warn, or raise in strict mode) from
  'found but already frozen' (recorded with a zero count, not an error)

Tests pin the parity: a bare name does not match nested submodules,
strict mode raises on unresolved paths, and the stage produces the same
requires_grad map as freeze_submodule_by_name for the same target list.

anaprietonem commented on training/docs/user-guide/training.rst:
```diff
@@ -562,39 +562,55 @@ The above can be adapted depending on the use case and taking advantage
+      model_modifier:
+         modifiers:
+            - _target_: "anemoi.training.checkpoint.modifiers.freezing.FreezingModifierStage"
+              submodules_to_freeze: ["encoder", "processor.0"]
```
@JPXKQX since you mentioned that you recently spot an issue with freezing, can you confirm this fits fine with the test you were doing in the multi-dataset set up?

anaprietonem commented on training/src/anemoi/training/checkpoint/modifiers/freezing.py:
```diff
@@ -0,0 +1,207 @@
+        frozen_count = 0
+        for param in target_module.parameters():
+            if param.requires_grad:
+                param.requires_grad = False
```
guess you could also do here something like?
```
frozen_count = sum(p.requires_grad for p in target_module.parameters())
target_module.requires_grad_(False)
```

JesperDramsch replied:
Good point, done. Swapped to `sum(p.requires_grad ...)` + `requires_grad_(False)`, same behaviour just less code.

anaprietonem commented on training/src/anemoi/training/checkpoint/modifiers/freezing.py:
```diff
@@ -0,0 +1,207 @@
+                frozen_count += 1
+        return frozen_count
+
+    def _validate_gradient_flow(self, model: torch.nn.Module) -> None:
```
do you have a feeling for how expensive is to do this check?

JesperDramsch replied:
Honestly negligible. It was a one-off forward/backward at load, but the dummy `(1,10)` input doesn't match real models so it just warned and skipped there anyway. Swapped it for a plain `requires_grad` check on the frozen submodules, which is free, and `requires_grad=False` already guarantees no grads.

anaprietonem commented on training/src/anemoi/training/config/training/single.yaml:
```diff
@@ -4,6 +4,7 @@ defaults:
   - training_loss: single
   - optimization: default
   - weight_averaging: null
+  - optional model_modifier: null
```
do we need this now? As in if the integration work will come after we could just leave the configs and then update all the other top level configs in the final integration PR

JesperDramsch replied:
That's fair. I was leaning towards "keeping it moving" rather than wait. Pushing to Phase 3

anaprietonem commented on training/docs/user-guide/training.rst:
```diff
@@ -562,39 +562,55 @@ The above can be adapted depending on the use case and taking advantage
 ``config.training.fork_run_id`` to define the path to the checkpoint.
 
-*******************
- Transfer Learning
```
Don't think the section about transfer learning should be removed.

JesperDramsch replied:
Agreed, reverted the user-guide back to `main` so the Transfer Learning section is restored.

anaprietonem commented on training/docs/user-guide/training.rst:
```diff
@@ -562,39 +562,55 @@ The above can be adapted depending on the use case and taking advantage
+              strict: false
+              validate_gradients: true
+
+Legacy key
```
Similarly, the comment below. At the moment the modifiers can't be used till we have the integration PR, so I think it would be better to change this page then so we simply update it and no need to keep any 'legacy entries'

JesperDramsch replied:
Yep, makes sense. Dropped the modifier docs here, I'll redo this page in the integration PR once they're actually usable. No legacy entries.

anaprietonem reviewed (commented):
Nice work @JesperDramsch ! thanks for the PR. Left some minor suggestions. Happy to have a another look after!

JesperDramsch committed:
fix(training): remove orphaned checkpoint_loading config templates

The four loader templates (weights_only, transfer_learning, warm_start, cold_start) under config/training/checkpoint_loading/ were added alongside the modifier work but belong to the config-wiring layer, not the modifier PR. Nothing references or instantiates them on this branch.

The canonical source and loading config group now lives at config/training/checkpoint/{source,loading}/ (PR #1196). transfer_learning.yaml also carried an unsupported strict key that raises TypeError when instantiated. config/training/model_modifier/freezing.yml is unchanged.

JesperDramsch committed:
refactor(training): move model modifiers under training.checkpoint namespace

Unify the checkpoint pipeline configuration under one namespace: model modifiers now live at training.checkpoint.modifiers (an ordered list), alongside training.checkpoint.source and training.checkpoint.loading, instead of the separate top-level training.model_modifier.

Moves config/training/model_modifier/freezing.yml to config/training/checkpoint/modifier/freezing.yml; removes the model_modifier schema field (superseded by the training.checkpoint field in PR #1196) and the model_modifier default group from single.yaml; updates the user-guide examples to the new namespace.

JesperDramsch committed:
docs: restore transfer-learning guide; defer model-modifier usage docs to integration PR

Per review on #442: the modifier pipeline is not usable until the integration PR wires it, so its usage docs (and the Transfer Learning section they replaced) are reverted to main and will be (re)written in the integration PR. Restores the Transfer Learning user-guide section.

JesperDramsch committed:
refactor: rename checkpoint/modifier -> checkpoint/modifiers to match namespace key

The builder reads the ordered list at training.checkpoint.modifiers (plural); the config group dir now matches that key. Content unchanged.

JesperDramsch committed:
refactor(training): simplify submodule freezing per review

Adopt the concise frozen-count + requires_grad_(False) form from review. Behavior is unchanged: nn.Module.parameters() and nn.Module.requires_grad_() both recurse into submodules, so all parameters under the target are frozen and the count covers only previously-trainable params.

JesperDramsch committed:
refactor(training): validate freezing via requires_grad flags, not a forward pass

The gradient-flow check ran a forward/backward on a hardcoded (1,10) dummy input, which does not fit real Anemoi models — it just warned and skipped there. requires_grad=False already guarantees a parameter cannot accumulate gradients, so assert that flag directly on the frozen submodules instead: free, model-agnostic, and always runs. Drops the unused _check_module_gradients helper; torch moves to a TYPE_CHECKING import.

JesperDramsch committed:
Merge remote-tracking branch 'origin/main' into 410-refactor-of-model-initialisation-ie-weight-loading-model-freezing-transfer-learning

JesperDramsch committed:
test(training): cover #1159 freezing parity scenarios in modifier tests

Add FreezingModifierStage ports of the two legacy freeze_submodule_by_name
scenarios from #1159 (training/tests/unit/utils/test_checkpoint_freezing.py):
a deep dot-path (a.sequential.0) freezing only that submodule, and a bare
branch name (a) freezing the whole subtree, both on a two-branch ModuleDict
that checks the sibling branch stays trainable. Keeps the legacy util test
beside its util while giving the modifier standalone coverage of the same
path semantics.

anaprietonem reviewed (approved):
Thanks for the contribution Jesper! and for the addressing the comments, good to be merged!

JesperDramsch committed:
Merge remote-tracking branch 'origin/main' into 410-refactor-of-model-initialisation-ie-weight-loading-model-freezing-transfer-learning

JesperDramsch committed:
Merge branch 'main' into 410-refactor-of-model-initialisation-ie-weight-loading-model-freezing-transfer-learning
